### PR TITLE
Introduce scope to semantic analyzer and update correlation in Apply node

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
@martint  To make this PR be easier to review I separated refactorings into separate review (https://github.com/prestodb/presto/pull/5303). That way it will be easier to merge it partially. 

The most important commit is https://github.com/prestodb/presto/pull/5302/commits/ab27e14e41c0eb236a606e6b7323dea189319dee

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -1133,21 +1078,25 @@ private static ExpressionAnalysis analyzeExpressions(
             Metadata metadata,
             SqlParser sqlParser,
             RelationType tupleDescriptor,
-            Iterable<? extends Expression> expressions)
+            Iterable<Expression> expressions)
     {
         // expressions at this point can not have sub queries so deny all access checks
         // in the future, we will need a full access controller here to verify access to functions
-        ExpressionAnalyzer analyzer = create(new Analysis(), session, metadata, sqlParser, new DenyAllAccessControl(), false);
+        Node root = Iterables.size(expressions) > 0 ? ExpressionUtils.and(expressions) : TRUE_LITERAL;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Why do we create a boolean expression?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        this.parent = requireNonNull(parent, "parent is null");
+        this.relation = requireNonNull(relation, "relation is null");
+    }
+
+    public Scope()
+    {
+        this(Optional.empty(), new RelationType());
+    }
+
+    public RelationType getRelation()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Rename this to `getType` or `getRelationType`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticExceptions.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.AMBIGUOUS_ATTRIBUTE;
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.MISSING_ATTRIBUTE;
+
+public final class SemanticExceptions
+{
+    private SemanticExceptions() {}
+
+    public static SemanticException createMissingAttributeException(Expression node)
+    {
+        return new SemanticException(MISSING_ATTRIBUTE, node, "Column '%s' cannot be resolved", node);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticExceptions.java]: It's not a good idea to rely on the node's toString for printing the column name.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticExceptions.java]: Oh, you just moved this code... nevermind, then (but we should fix this at some point)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: When somebody calls `Analysis.getScope(node)` and such `node` has not have mapping to scope. Then to find its scope a visitor is used, it finds the nearest parent node of this `node` which has a scope set and returns it as a `node` scope. This search is started from the root node.

Here I wanted to create such `artificial` root for all the expressions. But now, when you have asked I realized that is not needed and what is worse it introduces unnecessary confusion. So I removed that. Instead I create multiple roots, one per each expression.

**Comment by kokosing** [comment]
@martint Comments applied and rebased.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Sets.newIdentityHashSet;
+import static java.util.Objects.requireNonNull;
+
+public class Scope
+{
+    private final Optional<Scope> parent;
+
+    private RelationType relation;
+    private Optional<Scope> lateralScope = Optional.empty();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Why do we need to call out the lateral scope explicitly? Unless I'm mistaken, when lateral is used, the lateral derived table is behaves almost like a subquery, and the relation on the left of it acts as the "parent" scope.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ScopeFactory.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.Node;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ScopeFactory
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ScopeFactory.java]: Do we really need this class? Can't we make "Scope" have the ability to create the child scopes? (just brainstorming... I haven't thought deeply about it)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.DUPLICATE_RELATION;
+import static com.facebook.presto.sql.analyzer.SemanticExceptions.createAmbiguousAttributeException;
+import static com.facebook.presto.sql.analyzer.SemanticExceptions.createMissingAttributeException;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Sets.newIdentityHashSet;
+import static java.util.Objects.requireNonNull;
+
+public class Scope
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: I don't like this class being mutable. Scopes are pretty static (unlike the Analysis, which needs to accumulate things as the analyzer walks the query).

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 7]
```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
...
+        private Optional<Scope> result = Optional.empty();
+
+        public GetScopeVisitor(IdentityHashMap<Node, Scope> scopes, Node node)
+        {
+            this.scopes = requireNonNull(scopes, "scopes is null");
+            this.node = requireNonNull(node, "node is null");
+        }
+
+        @Override
+        public Void process(Node current, LinkedList<Scope> stack)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Why do we need a stack? Wouldn't it be sufficient to just pass the "candidate" scope as the tree is walked recursively (i.e., leverage the call stack, instead)

Something like:

```java
Void process(Node current, Scope candidate)
{
    if (result.isPresent()) {
        return null;
    }

    Scope newCandidate = scopes.get(current);
    if (newCandidate == null) {
         newCandidate = candidate;
    }

    if (node == current) {
        result = Optional.of(newCandidate);
        return null;
    }

    return super.process(current, newCandidate);
}
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 8]
```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
...
+
+        if (root.isPresent()) {
+            GetScopeVisitor visitor = new GetScopeVisitor(scopes, node);
+            visitor.process(root.get(), new LinkedList<>());
+            return visitor.getResult();
+        }
+        return Optional.empty();
+    }
+
+    public Node getRoot()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Is this the root of the AST? What does it mean to have multiple roots? What's special about this one being on the first position of the `roots` list?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 9]
```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
...
+        return roots.get(0);
     }
 
-    public void setOutputDescriptor(Node node, RelationType descriptor)
+    public Scope getRootScope()
+    {
+        return getScope(getRoot());
+    }
+
+    private static class GetScopeVisitor
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: One thing that bugs me about this is that it assumes there's a link between lexical structure, AST structure and scoping rules. One case where this doesn't hold is lateral joins, for instance (per my other comment). I can imagine other scenarios in the future where this may not be true, such as a sequences of statements, procedures, etc.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Your suggestion has worked. I used the "parent" scope instead.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ScopeFactory.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ScopeFactory.java]: I removed it. I had it to automatically add `Scope` to `Analysis` when it is created. But it was not nice. I didn't like it either. I created a `Scope.Builder` instead and added `createScope` methods to `StatementAnalyzer`. Which is much better, although the builder usage is quite verbose.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: I wanted to make immutable, but it was not convenient for all the use cases. Now with `Builder` is it is better.

Now, once you have asked, I made it immutable. To achieve that I had to move the field `columnReferences` to `Analysis` (anyway it is a better place for that). 

Take a look `StatementAnalyzer#visitQuery` and `StatementAnalyzer#analyzeWith` methods. These are the only places where immutability of `Scopes` makes troubles, because I need to create partial `Scope` objects to analyze nested nodes. To me it is acceptable, but I would like to hear your voice about it.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: yeah... it was hacky. It come from the fact `StatementAnalyzer` were doing a query rewrites. I wanted to have one `root`, which I could traverse when looking for a `Scope`. But in case of a rewrite it become tricky. I could lose information about old statement `root` by overwriting the `root` or I could have multiple roots. Both solutions were bad.

I extracted rewrites from `StatementAnalyzer`. Now I have one `root`, `StatementAnalyzer` is more specialized (focused) and easier to understand.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Such cases have to be treated separately. Each of such statement (or procedure) have to get special parent scope which will contain information produced by previous statements (or procedures). Such cases already exists in the code. See `StatementAnalyzer#visitQuerySpecyfication` or `StatementAnalyzer#analyzeWith`. 

QuerySpecyfication can be treated as sequence of following "statements": 
 - scan (FROM clause)
 - filter (WHERE cluase)
 - aggregation 
 - projection
Each of above is manipulating on scope produced by previous statement. In code it is simplified (scan and filter uses one `sourceScope`  as well as aggregation and projection uses one `outputScope`).

Lets take a look how `StatementAnalyzer#analyzeWith` works.
```
- main query
 - with
   - a - query1
   - b - query2
   - c - query3
```

In above example when analyzing `query1`, `query2` and `query3` each of the queries gets different parent scope:
 - `query1` gets original parent scope
 - `query2` gets original parent scope with results from `query1`
 - `query3` gets parent scope of `query2` with added results of `query2`.
Then to `with` node I bound the `scope`  which contains results from all the queries (parent scope of `query3` with added results of `query3`).

**Comment by kokosing** [comment]
@martint All the comments applied and rebased. Does spliting pull requests help you reviewing? If so I can extract query rewites as a separate pull request.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+    }
+
+    private static class Visitor
+            extends AstVisitor<Node, Void>
+    {
+        private final Metadata metadata;
+        private final Session session;
+        private final SqlParser sqlParser;
+
+        public Visitor(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Put args on the same line

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 11]
```diff
@@ -402,51 +394,16 @@ private static void validateColumns(Statement node, RelationType descriptor)
...
-
-    private String getQueryPlan(Explain node, ExplainType.Type planType, ExplainFormat.Type planFormat)
-            throws IllegalArgumentException
-    {
-        switch (planFormat) {
-            case GRAPHVIZ:
-                return queryExplainer.get().getGraphvizPlan(session, node.getStatement(), planType);
-            case TEXT:
-                return queryExplainer.get().getPlan(session, node.getStatement(), planType);
+        checkState(node.isAnalyze(), "Non analyze explain should be rerwitten to Query");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: typo: "rerwitten"

**Comment by kokosing** [comment]
@martint rebased, comments applied, ping :)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 12]
```diff
@@ -384,50 +385,53 @@ private static void validateColumns(Statement node, RelationType descriptor)
...
+        Scope withScope = analyzeWith(node, scope);
+        boolean approximate = isApproximate(node);
+        Scope queryScope = Scope.builder().withParent(withScope).withApproximate(approximate).build();
+        Scope queryBodyScope = process(node.getQueryBody(), queryScope);
+        analyzeOrderBy(node, queryBodyScope);
+
+        // Input fields == Output fields
+        analysis.setOutputExpressions(node, descriptorToFields(queryBodyScope));
+
+        queryScope = Scope.builder().withParent(withScope).withRelationType(queryBodyScope.getRelationType()).withApproximate(approximate).build();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: put each `.withXXX` in a separate line for readability.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 13]
```diff
@@ -700,62 +689,57 @@ protected RelationType visitSetOperation(SetOperation node, AnalysisContext cont
...
         if (criteria instanceof NaturalJoin) {
             throw new SemanticException(NOT_SUPPORTED, node, "Natural join not supported");
         }
 
-        AnalysisContext leftContext = new AnalysisContext(context, new RelationType());
-        RelationType left = process(node.getLeft(), context);
-        leftContext.setLateralTupleDescriptor(left);
-        RelationType right = process(node.getRight(), leftContext);
+        Scope left = process(node.getLeft(), scope);
+        Scope right = process(node.getRight(), left);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Using `left` as the scope for `right` is only valid for lateral joins, no?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 14]
```diff
@@ -1560,12 +1537,19 @@ private void analyzeWith(Query node, AnalysisContext context)
                     throw new SemanticException(MISMATCHED_COLUMN_ALIASES, withQuery, "WITH column alias list has %s entries but WITH query(%s) has %s columns", columnNames.size(), name, queryDescriptor.getVisibleFieldCount());
                 }
             }
+            if (withScopeBuilder.containsNamedQuery(name)) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Why did this move below?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 15]
```diff
@@ -202,13 +203,14 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
         }
         ProjectNode projectNode = new ProjectNode(idAllocator.getNextId(), plan.getRoot(), assignments.build());
 
-        RelationType tupleDescriptor = new RelationType(visibleTableColumns.stream()
+        ImmutableList<Field> fields = visibleTableColumns.stream()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: Use `List` as the declared type for the variable

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java [hunk 16]
```diff
@@ -52,9 +55,21 @@ public RelationPlan(PlanNode root, RelationType descriptor, List<Symbol> outputS
...
     }
 
+    public Optional<Symbol> getSymbol(Expression expression)
+    {
+        Optional<ResolvedField> resolvedField = scope.tryResolveField(expression);
+        if (!resolvedField.isPresent()) {
+            return Optional.empty();
+        }
+
+        checkState(resolvedField.isPresent() && resolvedField.get().isLocal(), "Unable to get symbol for outer reference: '%s'", expression);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java]: Maybe rephrase as "Unable to get symbol for field in outer scope"

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 17]
```diff
@@ -4097,13 +4097,19 @@ public void testWithHiding()
             throws Exception
     {
         assertQuery("" +
-                "WITH a AS (SELECT custkey FROM orders), " +
+                "WITH a AS (VALUES 1), " +
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 18]
```diff
@@ -19,17 +19,28 @@
 import com.facebook.presto.sql.planner.plan.ApplyNode;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: This commit needs an explanation in the commit message.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 19]
```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
+            for (Symbol symbol : node.getOutputSymbols()) {
+                missing.remove(symbol.toQualifiedNameReference());
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
+        {
+            ImmutableSet<Expression> columnReferences = node.getAssignments().values().stream()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Use `Set` as the declared type for this variable

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 20]
```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
+        return visitor.getMissing();
+    }
+
+    private static class MissingReferencesExtractor
+            extends PlanVisitor<Void, Void>
+    {
+        private final Set<Expression> missing = new HashSet<>();
+        private final Set<Expression> columnReferences;
+
+        private MissingReferencesExtractor(Analysis analysis)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Pass the column references directly instead of the Analysis object.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 21]
```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
+            columnReferences = analysis.getColumnReferences();
+        }
+
+        @Override
+        protected Void visitPlan(PlanNode node, Void context)
+        {
+            node.getSources().forEach(source -> source.accept(this, null));
+
+            for (Symbol symbol : node.getOutputSymbols()) {
+                missing.remove(symbol.toQualifiedNameReference());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: This doesn't look right. There's no relationship between the names of the outputs and the column references in the analysis.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 22]
```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
+    }
+
+    private Set<Expression> extractMissingExpressions(PlanNode planNode)
+    {
+        MissingReferencesExtractor visitor = new MissingReferencesExtractor(analysis);
+        planNode.accept(visitor, null);
+        return visitor.getMissing();
+    }
+
+    private static class MissingReferencesExtractor
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Not sure I understand what this is trying to do.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 23]
```diff
@@ -127,30 +140,176 @@ private PlanBuilder appendScalarSubqueryApplyNode(PlanBuilder subPlan, SubqueryE
...
+            node.getSources().forEach(source -> source.accept(this, null));
+
+            for (Symbol symbol : node.getOutputSymbols()) {
+                missing.remove(symbol.toQualifiedNameReference());
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: What about other node types (filter, aggregation, sort, etc)?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 17 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: I will extract this as separate commit. I was using this tests in regards of scope of named query. 

IMO test was wrong as result of the query was the same disregarding which named query was executed: `a` or `b::a`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java [hunk 24]
```diff
@@ -60,32 +60,20 @@ public TestLogicalPlanner()
     }
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: I'm not sure I understand how these changes fix the issue described in the commit message. It seems like a minor change in the organization of the tests.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java [hunk 25]
```diff
@@ -33,6 +33,7 @@
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: Can you add a test that reproduces this problem?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 26]
```diff
@@ -69,6 +69,11 @@
 
 public class LogicalPlanner
 {
+    public enum Stage
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: What's this for? It doesn't seem to have any effect, since the only value that's ever used is OPTIMIZED_AND_VALIDATED.

**Comment by kokosing** [comment]
I extracted `StatementAnalyzer` refactoring to https://github.com/prestodb/presto/pull/5585. So it should be easier to review.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 14 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I am not sure if understand question correctly. Anyway I will try to answer it :)

`Scope.Builder` does have a knowledge about the AST nodes, throwing `SemanticException` from it would be strange. So this is why I have to check if query is not already added to builder before adding it there.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I meant that you moved this check from above, in line 1551, to here.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Ah... good catch. Thanks.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 21 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Now, with `SymbolReference` it should be easy to find column reference.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java (hunk 24 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: Well, there is no issue ;)

I just removed tests whose the main purpose was to check if logical plan assertions are working. Instead I left test whose the main purpose is to check if logical planner return correct plan. 

Logical plan assertions are tested anyway, I do think we need special tests for that.

As you said it is a minor change in the organization of the tests ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 26 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: It is for tests ; /

Currently when plan is validated and find correlation (non rewritten Apply) will raise an exception, but in tests I want to get the plan before validation to check if correlation was resolved correctly.

**Comment by kokosing** [comment]
@martint Rebased and applied all the comments. Also, please notice that I extracted some refactoring commits as a separate review.

**Comment by kokosing** [comment]
Travis errored with same failures as for master branch.

**Comment by martint** [comment]
@kokosing, I think there are some comments/questions that still need to be addressed. Can you check?

Also, this doesn't rebase cleanly after I merged the EXISTS and statement rewriter PRs.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 12 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: `left` is also valid as parent scope for `JOIN UNNEST`. I was unable to model that differently than restoring `lateralScope` field in `Scope` class.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 22 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: I removed that class

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java (hunk 25 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: This code got removed.

**Comment by kokosing** [comment]
@martint Sorry about missing comments. I had to do something wrong with the git and I lost some changes.

Please notice that I restored the `lateralScope` in the `Scope` class. I had no better idea how to handle `JOIN UNNEST`. Moreover I think it will be needed for `LATERAL JOIN`. If you have a better idea then I will be glad to change that.

This pull request still need some work. I need to add unit tests for correlated `EXISTS` subqueries, because queries for correlated `EXISTS` we had so far are passing as it was possible to prune correlation and in some cases whole subquery was eligible to be pruned.

Anyway, it would be good to merge some of the commits as this pull request is getting bigger and bigger. I have heard that you are going on vacation so I don't if it is possible in near future.

**Comment by kokosing** [comment]
I split this pull request into two smaller ones.

Superseded by https://github.com/prestodb/presto/pull/5686 and https://github.com/prestodb/presto/pull/5687

**Comment by martint** [comment]
> Please notice that I restored the lateralScope in the Scope class. I had no better idea how to handle JOIN UNNEST. Moreover I think it will be needed for LATERAL JOIN. If you have a better idea then I will be glad to change that.

Why can't we model it as:

* `A JOIN B`: A and B have the outer scope as parents
* `A JOIN LATERAL B`: A has outer scope as parent, B has A's scope as parent
* `A, UNNEST(...)`: A has outer scope as parent, UNNEST has A's scope as parent

Semantically, UNNEST implies "CROSS JOIN LATERAL", so I don't see why we'd treat it any different from common LATERAL.

**Comment by kokosing** [comment]
According to my understanding, what you have proposed is the semantic of that SQL constructs. It it is the only way it can be modeled to make this construct work correctly. Moreover, in this PR it scopes are implemented that way, but maybe it is not obvious because of the usage of `lateralScope`.

I will try to remove `lateralScope` by checking explicitly in `StatementAnalzyer#visitJoin` the relation type of the right side of the join.

**Comment by kokosing** [comment]
It worked, please https://github.com/prestodb/presto/pull/5686 and its recent commit: https://github.com/prestodb/presto/pull/5686/commits/f7764d21817083cfe097900b2bcd05d8c5756a7e
